### ISSUE TITLE
RTDetrV2ForObjectDetection produces different outputs in Transformers v5 with identical inputs

### ISSUE THREAD

**Issue by MorganFujimaka** [huggingface/transformers#43697]
### System Info

After upgrading from Transformers 4.57.6 to 5.0.0, RTDetrV2ForObjectDetection produces different logits and pred_boxes for identical pixel_values tensors.
	•	The same saved pixel_values tensor is reused across versions (`torch.randn(1, 3, 640, 640, dtype=torch.float32)`).
	•	Model is in eval() mode
	•	Weights/config are identical (checkpoint trained with Transformers 4.51.3).
	•	Differences are non-trivial, not just numeric noise:
```
# transformers 4.57.6 outputs.pred_boxes
tensor([[[0.0803, 0.1549, 0.0167, 0.0167],
         [0.3277, 0.0824, 0.0288, 0.0235],
         [0.5289, 0.8680, 0.0158, 0.0149],
         ...,
         [0.1689, 0.2689, 0.0320, 0.0193],
         [0.4674, 0.4583, 0.0376, 0.0227],
         [0.7807, 0.9875, 0.0544, 0.0248]]])

# transformers 5.0.0 outputs.pred_boxes
tensor([[[0.0789, 0.1558, 0.0198, 0.0225],
         [0.3244, 0.0851, 0.0417, 0.0465],
         [0.5289, 0.8686, 0.0192, 0.0195],
         ...,
         [0.1615, 0.2735, 0.0409, 0.0342],
         [0.4569, 0.4622, 0.0590, 0.0432],
         [0.7627, 0.9902, 0.0575, 0.0225]]])
```

This suggests a behavioral change in the RTDetrV2ForObjectDetection forward implementation between Transformers 4.x and 5.0.0.

Happy to provide a minimal reproduction with a fixed input tensor and checkpoint if helpful.
[config.json](https://github.com/user-attachments/files/25035123/config.json)

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import RTDetrV2ForObjectDetection

pixel_values = torch.randn(1, 3, 640, 640, dtype=torch.float32)
torch.save(pixel_values, "test_pixel_values.pt")

# Switch to transformers 4.57.6
model = RTDetrV2ForObjectDetection.from_pretrained(model_path)
pred_boxes_v4 = model(pixel_values=pixel_values).pred_boxes

# Switch to transformers 5.0.0
pixel_values = torch.load("test_pixel_values.pt")

model = RTDetrV2ForObjectDetection.from_pretrained(model_path)
pred_boxes_v5 = model(pixel_values=pixel_values).pred_boxes

torch.equal(pred_boxes_v4, pred_boxes_v5) => False
```

### Expected behavior

With identical pixel_values and the same checkpoint, RTDetrV2ForObjectDetection should produce identical (or numerically equivalent) outputs across Transformers versions.

**Comment by NielsRogge** [comment]
Hi @MorganFujimaka can you check whether my PR above fixes the issue?

**Comment by yonigozlan** [comment]
Hello @MorganFujimaka ! It'd be great if you could provide a checkpoint to reproduce the issue. @NielsRogge I don't think the issue comes form sin pos embeddings. The ones in v5 should actually be correct for non square image inputs (see https://github.com/huggingface/transformers/pull/41380), and they are equivalent to the old ones for square image inputs (rt_detr_v2 checkpoints on the hub resize the image to squares so this should make no differences).
What could be different is that we switched to using fast image processors (torch/torchvision based) by default which results in very slight numerical differences in processed images outputs, which can be amplified by the forward pass. Can you try to check if you get the old results back by instantiating the processor using `use_fast=False`? Otherwise can you provide the checkpoints you're using to reproduce?
Out of curiosity, do the new results make sense or are worse than the old ones?

**Comment by MorganFujimaka** [comment]
@yonigozlan Looks like it’s resolved in v5.1.0. Answering your questions:
1. Switching to `use_fast=False` didn’t help.
2. I haven’t tested it extensively, only a couple of test cases failed. The updated bounding boxes still looked reasonable, just slightly different, which was enough to break our tests. That instability made me hesitant to roll out Transformers 5.0.0 to production.
3. For reproduction, I used a model fine-tuned with Transformers 4.51.3. It behaved as expected on 4.57.6, but started failing tests after upgrading to 5.0.0.

**Comment by NielsRogge** [comment]
Hi, let me take a look at this one.

**Comment by NielsRogge** [comment]
Hi @MorganFujimaka, I had Codex investigate this issue:

Key result: the v5.0.0 regression is caused by RT-DETR v2 _tied_weights_keys mapping direction, which prevents v4-style model.decoder.* head weights from loading correctly (even though missing/unexpected key counters stay zero). This was corrected by #41549 (present in v5.1.0), and outputs return to near-identical parity with v4.57.6.
